File: templates/problems/08_tries/01_trie/solution.go
package trie

// solveTrieNode represents a node in the SolveTrie.
type solveTrieNode struct {
	Children map[byte]*solveTrieNode
	IsEnd    bool
}

// SolveTrie implements a prefix tree.
type SolveTrie struct {
	Root *solveTrieNode
}

// NewSolveTrie creates and returns a new SolveTrie.
func NewSolveTrie() SolveTrie {
	return SolveTrie{Root: &solveTrieNode{Children: make(map[byte]*solveTrieNode)}}
}

// Insert adds a word to the trie.
// Time: O(m) where m is word length, Space: O(m).
func (t *SolveTrie) Insert(word string) {
	node := t.Root
	for i := 0; i < len(word); i++ {
		c := word[i]
		if node.Children[c] == nil {
			node.Children[c] = &solveTrieNode{Children: make(map[byte]*solveTrieNode)}
		}
		node = node.Children[c]
	}
	node.IsEnd = true
}

// Search returns true if the word is in the trie.
// Time: O(m), Space: O(1).
func (t *SolveTrie) Search(word string) bool {
	node := t.findNode(word)
	return node != nil && node.IsEnd
}

// StartsWith returns true if any word in the trie starts with the given prefix.
// Time: O(m), Space: O(1).
func (t *SolveTrie) StartsWith(prefix string) bool {
	return t.findNode(prefix) != nil
}

func (t *SolveTrie) findNode(prefix string) *solveTrieNode {
	node := t.Root
	for i := 0; i < len(prefix); i++ {
		node = node.Children[prefix[i]]
		if node == nil {
			return nil
		}
	}
	return node
}

File: templates/problems/08_tries/01_trie/python/solution.py
from __future__ import annotations


class _SolveTrieNode:
    __slots__ = ("children", "is_end")

    def __init__(self):
        self.children: dict[str, _SolveTrieNode] = {}
        self.is_end = False


class SolveTrie:
    """A prefix tree. Time per op: O(m) where m is word length."""

    def __init__(self):
        self.root = _SolveTrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = _SolveTrieNode()
            node = node.children[c]
        node.is_end = True

    def search(self, word: str) -> bool:
        node = self._find_node(word)
        return node is not None and node.is_end

    def starts_with(self, prefix: str) -> bool:
        return self._find_node(prefix) is not None

    def _find_node(self, prefix: str) -> _SolveTrieNode | None:
        node = self.root
        for c in prefix:
            node = node.children.get(c)
            if node is None:
                return None
        return node
